### PR TITLE
Add llama end to end test with jetstream engine 

### PR DESCRIPTION
**Pull request by FanhaiLu1**
The purpose of this PR is testing the E2E accuracy between Jetstream engine inference and original llama inference. Below are tests:

- test_llama_e2e.  original llama inference 10 steps decode **vs** jetstream engine 10 steps
- test_original_llama2_seed.   original llama inference 10 steps with different seed
- test_jetstream_llama2_seed. jetstream engine 10 with different seed
- test_llama_e2e_two_addtional_tokens.  Same prefix as test_llama_e2e, add two additional tokens in prompts.  
-  test_llama_e2e_four_addtional_tokens.  Same prefix as test_llama_e2e, add four additional tokens in prompts.

### CODE REVIEW

File: jetstream_pt/third_party/llama2/generation_original.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+import json
+import os
+import sys
+import time
+from pathlib import Path
+from typing import List, Literal, Optional, Tuple, TypedDict
+
+import torch
+import torch.nn.functional as F
+from fairscale.nn.model_parallel.initialize import (
```
Comment (qihqi) [inline jetstream_pt/third_party/llama2/generation_original.py:13]: these imports seems unused.

**Review by qihqi — APPROVED** [review]

File: tests/test_llama_e2e.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+        for index, output_tokens in enumerate(output_tokens_multiple): 
+            print(f"------------------- index: {index}, tokens:{output_tokens}")
+            if index > 0:
+                self.assertNotEqual(output_tokens_multiple[index], output_tokens_multiple[index - 1])
+
+    def test_jetstream_llama2_seed(self):
+        jax.config.update('jax_platform_name', 'cpu')
+        x = jnp.square(2)
+        print(f"---------> {repr(x.addressable_data(0).devices())}")
```
Comment (wang2yn84) [inline tests/test_llama_e2e.py:70]: Consider jax.devices()?

File: tests/test_llama_e2e.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+        for index, output_tokens in enumerate(output_tokens_multiple): 
+            print(f"------------------- index: {index}, tokens:{output_tokens}")
+            if index > 0:
+                self.assertEqual(output_tokens_multiple[index], output_tokens_multiple[index - 1])
+
+    def test_llama_e2e(self):
+        jax.config.update('jax_platform_name', 'cpu')
+        x = jnp.square(2)
+        print(f"---------> {repr(x.addressable_data(0).devices())}")
```
Comment (wang2yn84) [inline tests/test_llama_e2e.py:137]: Ditto

File: tests/test_llama_e2e.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+                break
+        print(f"-------------------->out_tokens:{out_tokens}")
+        print(f"-------------------->expected_output_tokens:{expected_output_tokens}")
+        # self.assertEqual(out_tokens ,expected_output_tokens)
+
+
+    def test_llama_e2e_four_addtional_tokens(self):
+        jax.config.update('jax_platform_name', 'cpu')
+        x = jnp.square(2)
+        print(f"---------> {repr(x.addressable_data(0).devices())}")
```
Comment (wang2yn84) [inline tests/test_llama_e2e.py:274]: Ditto

File: tests/test_llama_e2e.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+        env_data.batch_size = 1
+        env = environment.JetEngineEnvironment(env_data)
+        env.apply_sharding = lambda *args, **kwargs: None  # don't shard on cpu
+        return env
+
+
+    def test_original_llama2_seed(self):
+        jax.config.update('jax_platform_name', 'cpu')
+        x = jnp.square(2)
+        print(f"---------> {repr(x.addressable_data(0).devices())}")
```
Comment (wang2yn84) [inline tests/test_llama_e2e.py:48]: Ditto

File: tests/test_llama_e2e.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+        return pytree.tree_map_only(
+            torch.Tensor,
+            torch_xla2.tensor.t2j, tree)    
+
+    def _make_env(self):
+        jax.config.update('jax_dynamic_shapes', False)
+        jax.config.update('jax_traceback_filtering', 'off')
+        env_data = environment.JetEngineEnvironmentData()
+        env_data.max_input_sequence_length = 128
+        env_data.max_input_sequence_length = 128
```
Comment (wang2yn84) [inline tests/test_llama_e2e.py:36]: Redundant

File: jetstream_pt/third_party/llama2/generation_original.py (hunk 1 above)
Comment (FanhaiLu1) [reply jetstream_pt/third_party/llama2/generation_original.py]: Thanks, updated.

File: tests/test_llama_e2e.py (hunk 2 above)
Comment (FanhaiLu1) [reply tests/test_llama_e2e.py]: Thanks, updated.

File: tests/test_llama_e2e.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+        llama_original = LlamaOriginal.build(tokenizer_path, model_arg, seed)
+        model_orig = llama_original.model
+
+        state_dict = dict(model_orig.state_dict())
+        state_dict['freqs_cis'] = model_orig.freqs_cis
+        params = self._to_jax(state_dict)
+
+        output_tokens_multiple = []
+        for i in [1, 2, 3]:
+            torch.manual_seed(1)
```
Comment (wang2yn84) [inline tests/test_llama_e2e.py:98]: Why fix the seed here?

File: tests/test_llama_e2e.py [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+        env = self._make_env()
+        model_arg = env._model_arg 
+        tokens = np.arange(10, dtype=np.int32)
+        tokens = np.append(tokens, [15050, 3503], axis=-1)
+        true_length = tokens.shape[-1]
+        padded_tokens = np.pad(tokens, (0, 6))
+        padded_tokens = jnp.array(padded_tokens)
+
+        seed = 1
+        torch.manual_seed(1)
```
Comment (wang2yn84) [inline tests/test_llama_e2e.py:218]: Seed

File: tests/test_llama_e2e.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+        torch.set_default_dtype(torch.bfloat16)
+        env = self._make_env()
+        model_arg = env._model_arg 
+        tokens = np.arange(10, dtype=np.int32)
+        true_length = tokens.shape[-1]
+        padded_tokens = np.pad(tokens, (0, 6))
+        padded_tokens = jnp.array(padded_tokens)
+
+        seed = 1
+        torch.manual_seed(1)
```
Comment (wang2yn84) [inline tests/test_llama_e2e.py:148]: seed

File: tests/test_llama_e2e.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+                    break
+
+            output_tokens_multiple.append(out_tokens)
+
+        for index, output_tokens in enumerate(output_tokens_multiple): 
+            print(f"------------------- index: {index}, tokens:{output_tokens}")
+            if index > 0:
+                self.assertEqual(output_tokens_multiple[index], output_tokens_multiple[index - 1])
+
+    def test_llama_e2e(self):
```
Comment (wang2yn84) [inline tests/test_llama_e2e.py:134]: This test seems the same as the previous one?

File: tests/test_llama_e2e.py [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+    def test_llama_e2e_two_addtional_tokens(self):
+        jax.config.update('jax_platform_name', 'cpu')
+        x = jnp.square(2)
+        print(f"---------> {repr(x.addressable_data(0).devices())}")
+
+        torch.set_default_dtype(torch.bfloat16)
+        env = self._make_env()
+        model_arg = env._model_arg 
+        tokens = np.arange(10, dtype=np.int32)
+        tokens = np.append(tokens, [15050, 3503], axis=-1)
```
Comment (wang2yn84) [inline tests/test_llama_e2e.py:212]: Curious, so the tokens from 0 to 9 can generate the same result?

**Review by wang2yn84 — COMMENTED** [review]
Thank you very much for the detailed tests. Leave some comments here.

File: tests/test_llama_e2e.py (hunk 11 above)
Comment (FanhaiLu1) [reply tests/test_llama_e2e.py:212]: yes, the output tokens are same from different run.

File: tests/test_llama_e2e.py (hunk 10 above)
Comment (FanhaiLu1) [reply tests/test_llama_e2e.py:134]: diff one. Previous is testing different seed, this one is compared llama original vs jetstream pytroch.

File: tests/test_llama_e2e.py (hunk 9 above)
Comment (FanhaiLu1) [reply tests/test_llama_e2e.py:148]: will have next version soon
